apps/src/code-studio/pd/workshop_dashboard/reports/foorm/results.jsx

```diff
@@ -0,0 +1,154 @@
+            if (!answer) {
+              return null;
+            }
+
+            if (
+              ['scale', 'singleSelect', 'multiSelect'].includes(
+                question['type']
+              )
+            ) {
+              // numRespondents will get either a value (for multiSelect) or undefined.
+              const numRespondents = answer.num_respondents;
+
+              // Make a copy of the answers without the num_respondents and other_answers fields.
+              const filteredAnswers = _.omit(answer, [
+                'num_respondents',
+                'other_answers'
+              ]);
+
+              let possibleAnswersMap = question['choices'];
+              uniqueKey++;
+              return (
+                <ChoiceResponses
+                  perFacilitator={section === 'facilitator'}
+                  numRespondents={numRespondents}
+                  question={question['title'] || questionId}
+                  answers={filteredAnswers}
+                  possibleAnswers={Object.keys(possibleAnswersMap)}
+                  possibleAnswersMap={possibleAnswersMap}
+                  key={uniqueKey}
+                  answerType={question['type']}
+                  otherText={question['other_text']}
+                  otherAnswers={answer['other_answers']}
+                />
+              );
+            } else if (question['type'] === 'matrix') {
+              // render choice response per question inside matrix
+              if (!question['rows']) {
+                return null;
+              }
+              return Object.keys(question['rows']).map(innerQuestionId => {
+                const innerAnswer = answer[innerQuestionId];
+                const numRespondents = answer.num_respondents;
+                let possibleAnswersMap = question['columns'];
+                let parsedQuestionName = `${question['title']} -> ${
+                  question['rows'][innerQuestionId]
+                }`;
+                uniqueKey++;
+                return (
+                  <ChoiceResponses
+                    perFacilitator={section === 'facilitator'}
+                    numRespondents={numRespondents}
+                    question={parsedQuestionName}
+                    answers={innerAnswer}
+                    possibleAnswers={Object.keys(possibleAnswersMap)}
+                    possibleAnswersMap={possibleAnswersMap}
+                    key={uniqueKey}
+                    answerType={'singleSelect'}
+                  />
+                );
+              });
+            } else if (question['type'] === 'text') {
+              uniqueKey++;
+              return (
+                <TextResponses
+                  question={question['title'] || questionId}
+                  answers={answer}
+                  key={uniqueKey}
+                />
+              );
+            }
+          })
+        );
+      })
+    );
+  }
+
+  renderResultsForSession(session) {
```

Comment:
I know we have precedent for this idea of `render*()` functions which return new content inside a `<div>`, in PLC at least, but I would really like us to move to a more traditional model in which each of these is a named React component.  Amongst other benefits, it means that when we inspect the React DOM, we see helpfully-named components in a readable hierarchy, as opposed to a stack of anonymous `<div>`s of ambiguous origins.

Reply:
👍 I will update this

Reply:
did some refactoring, let me know if what I did makes sense/needs improvement